refactor: Consolidate ERC165 supportsInterface logic in AA and Strategy contracts

adamgall opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/312

Fixes [ENG-1071](https://linear.app/decent-labs/issue/ENG-1071/consolidate-and-refine-supportsinterface-in-aa-and-strategy-contracts)

This commit refactors the ERC165 `supportsInterface` implementations across `DecentPaymasterV1`, `SmartAccountValidationV1`, `ERC4337VoterSupportV1`, and `StrategyV1`.

The main changes are:
-   `SmartAccountValidationV1.sol`: Removed its `supportsInterface` function and direct inheritance from `ERC165`. `ERC165` is now expected to be inherited by implementing contracts if needed, or its support is checked at a higher level.
-   `ERC4337VoterSupportV1.sol`: Removed its `supportsInterface` function. Interface support is now handled by `StrategyV1`.
-   `DecentPaymasterV1.sol`: Its `supportsInterface` function now directly checks for `IDecentPaymasterV1`, `ISmartAccountValidationV1`, and `IPaymaster`, then calls `super.supportsInterface()`. The explicit override list `(SmartAccountValidationV1, Version)` has been removed as `Version` (which inherits `ERC165`) is implicitly handled by `super` and `SmartAccountValidationV1`'s interface check is now direct.
-   `StrategyV1.sol`: Its `supportsInterface` function now directly checks for `IStrategyV1`, `IStrategyBaseV1`, `IERC4337VoterSupportV1`, and `ISmartAccountValidationV1`, then calls `super.supportsInterface()`. The explicit override list `(Version, ERC4337VoterSupportV1)` has been removed.

Test updates:
-   `SmartAccountValidationV1.test.ts`: Removed ERC165 tests as the contract no longer implements `supportsInterface` directly.
-   `ERC4337VoterSupportV1.test.ts`: Removed ERC165 tests for the same reason.
-   `StrategyV1.test.ts`: Added tests to confirm `StrategyV1` now reports support for `IERC4337VoterSupportV1` and `ISmartAccountValidationV1`.

This refactoring centralizes the `supportsInterface` logic in the more concrete, top-level contracts, simplifying the intermediate abstract contracts and ensuring that the final deployed contracts accurately report all interfaces they implement through their inheritance chain.

tbwebb22 reviewed (commented):
what's the reasoning behind the ERC-165 changes in this PR? I would have that the original pattern was more consistent with what we have across the codebase

adamgall committed:
refactor: Consolidate ERC165 supportsInterface logic in AA and Strategy contracts

This commit refactors the ERC165 `supportsInterface` implementations across `DecentPaymasterV1`, `SmartAccountValidationV1`, `ERC4337VoterSupportV1`, and `StrategyV1`.

The main changes are:
-   `SmartAccountValidationV1.sol`: Removed its `supportsInterface` function and direct inheritance from `ERC165`. `ERC165` is now expected to be inherited by implementing contracts if needed, or its support is checked at a higher level.
-   `ERC4337VoterSupportV1.sol`: Removed its `supportsInterface` function. Interface support is now handled by `StrategyV1`.
-   `DecentPaymasterV1.sol`: Its `supportsInterface` function now directly checks for `IDecentPaymasterV1`, `ISmartAccountValidationV1`, and `IPaymaster`, then calls `super.supportsInterface()`. The explicit override list `(SmartAccountValidationV1, Version)` has been removed as `Version` (which inherits `ERC165`) is implicitly handled by `super` and `SmartAccountValidationV1`'s interface check is now direct.
-   `StrategyV1.sol`: Its `supportsInterface` function now directly checks for `IStrategyV1`, `IStrategyBaseV1`, `IERC4337VoterSupportV1`, and `ISmartAccountValidationV1`, then calls `super.supportsInterface()`. The explicit override list `(Version, ERC4337VoterSupportV1)` has been removed.

Test updates:
-   `SmartAccountValidationV1.test.ts`: Removed ERC165 tests as the contract no longer implements `supportsInterface` directly.
-   `ERC4337VoterSupportV1.test.ts`: Removed ERC165 tests for the same reason.
-   `StrategyV1.test.ts`: Added tests to confirm `StrategyV1` now reports support for `IERC4337VoterSupportV1` and `ISmartAccountValidationV1`.

This refactoring centralizes the `supportsInterface` logic in the more concrete, top-level contracts, simplifying the intermediate abstract contracts and ensuring that the final deployed contracts accurately report all interfaces they implement through their inheritance chain.

adamgall commented:
> what's the reasoning behind the ERC-165 changes in this PR? I would have that the original pattern was more consistent with what we have across the codebase

I agree I would have preferred to not implement the changes in this PR.

The reason, however, shows up two PRs down the line (https://github.com/decentdao/decent-contracts/pull/316), when adding `ERC4337VoterSupportV1` into the inheritance chain of the FreezeVoting contracts. Without the changes in this PR, I was getting "cannot resolve linearization graph" errors.
